x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java

```diff
@@ -54,6 +55,19 @@ public final PhysicalOperation groupingPhysicalOperation(
         var aggregates = aggregateExec.aggregates();
 
         var sourceLayout = source.layout;
+        AggregatorMode aggregatorMode;
+
+        if (mode == AggregateExec.Mode.FINAL) {
+            aggregatorMode = AggregatorMode.FINAL;
+        } else if (mode == AggregateExec.Mode.PARTIAL) {
+            if (aggregateExec.child() instanceof ExchangeSourceExec) {// the reducer step at data node (local) level
+                aggregatorMode = AggregatorMode.INTERMEDIATE;
+            } else {
+                aggregatorMode = AggregatorMode.INITIAL;
+            }
+        } else {
+            aggregatorMode = AggregatorMode.INITIAL;
```

Comment:
Should this be `AggregatorMode.SINGLE` instead? Can we `assert false: ...` here since we shouldn't reach this point in our current code base.